cedar-policy-validator/src/cedar_schema/ast.rs

```diff
@@ -29,11 +29,53 @@ use smol_str::ToSmolStr;
 
 use crate::json_schema;
 
+use super::err::UserError;
+
 pub const BUILTIN_TYPES: [&str; 3] = ["Long", "String", "Bool"];
 
 pub(super) const CEDAR_NAMESPACE: &str = "__cedar";
 
-pub type Schema = Vec<Node<Namespace>>;
+/// A struct that can be annotated, e.g., entity types.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Annotated<T> {
+    /// The struct that's optionally annotated
+    pub data: T,
+    /// Annotations
+    pub annotations: Annotations,
+}
+
+pub type Schema = Vec<Annotated<Namespace>>;
+
+pub fn deduplicate_annotations<T>(
+    data: T,
+    annotations: Vec<Node<(Node<AnyId>, Option<Node<SmolStr>>)>>,
+) -> Result<Annotated<T>, UserError> {
+    let mut unique_annotations: BTreeMap<Node<AnyId>, Option<Node<SmolStr>>> = BTreeMap::new();
+    for annotation in annotations {
+        let (key, value) = annotation.node;
+        if let Some((old_key, _)) = unique_annotations.get_key_value(&key) {
+            return Err(UserError::DuplicateAnnotations(
+                key.node,
+                Node::with_source_loc((), old_key.loc.clone()),
+                Node::with_source_loc((), key.loc),
+            ));
+        } else {
+            unique_annotations.insert(key, value);
+        }
+    }
+    Ok(Annotated {
+        data,
+        annotations: unique_annotations
+            .into_iter()
+            .map(|(key, value)| {
+                (
+                    key.node,
+                    Annotation::with_optional_value(value.map(|n| n.node), None),
```

Comment:
is `None` correct here?  does `value` contain a source location we could propagate?

Reply:
Fixed.